File: solutions/studyplan/programmingskills/easy_283/Solution.java
package studyplan.programmingskills.easy_283;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Objects;

public class Solution {

    /*  My Solution
     *  Runtime: 117ms
     *  Memory: 47.79MB
     */
    public static void moveZeroes(int[] nums) {
        int count = 0;
        for (int i : nums) {
            if (i == 0) {
                count++;
            }
        }

        while (count > 0) {
            for (int x = 1; x < nums.length; x++) {
                if (nums[x - 1] == 0) {
                    int temp = nums[x - 1];
                    nums[x - 1] = nums[x];
                    nums[x] = temp;
                }
            }
            count--;
        }
    }

    /*  ChatGpt Solution
     *  Runtime: 3ms
     *  Memory: 47.82MB
     */
    public static void moveZeroes1(int[] nums) {
        int lastNonZero = 0;

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != 0) {
                if (i != lastNonZero) {
                    int temp = nums[lastNonZero];
                    nums[lastNonZero] = nums[i];
                    nums[i] = temp;
                }
                lastNonZero++;
            }
        }
    }


    public static void main(String[] args) {
        int[] nums = {0, 1, 0, 3, 12};
        moveZeroes(nums);
        System.out.println(Arrays.toString(nums));
    }

}

File: solutions/studyplan/programmingskills/easy_283/Solution.kt
package studyplan.programmingskills.easy_283

/*  My Solution
   *  Runtime: 87ms
   *  Memory: 50.04MB
   */
fun moveZeroes(nums: IntArray): Unit {
    var count = 0
    for (i in nums) {
        if (i == 0) {
            count++
        }
    }

    while (count > 0) {
        for (x in 1..<nums.size) {
            if (nums[x - 1] == 0) {
                val temp = nums[x - 1]
                nums[x - 1] = nums[x]
                nums[x] = temp
            }
        }
        count--
    }
}

/*  CHatGpt Solution
   *  Runtime: 3ms
   *  Memory: 52.09MB
   */
fun moveZeroes1(nums: IntArray): Unit {
    var lastNonZero = 0
    for (i in nums.indices) {
        if (nums[i] != 0) {
            if (i != lastNonZero) {
                val temp = nums[lastNonZero]
                nums[lastNonZero] = nums[i]
                nums[i] = temp
            }
            lastNonZero++
        }
    }
}

fun main() {
    val nums = intArrayOf(1, 2, 0, 4, 5, 6, 7, 8, 9, 10)
    moveZeroes(nums)
    println(nums.contentToString())
}